### DISCUSSION TITLE
WASM - Godot Engine scripting

### DISCUSSION THREAD

**Discussion by fire** [omigroup/omigroup#160] [Ideas]
@AdamFrisby

Godot Engine works on 5-10 platforms and has a scripting language called gdscript. One of the platforms is WASM. I wanted to investigate using jitasm for gdscript.

Since I wanted to try using gdscript for wasm, as a sandboxing layer, here are some questions.

Can we generate wasm in wasm and have the wasm called?

Here's a chat log with the asmjit community:

**kobalicek**: (Petr Kobalicek)
   you would be basically generating wasm code from wasm code, and wanting wasm code to call that generated wasm code
   I have no idea whether it's possible or not
**iFire**
    I have a feeling it is, but I'm new to this.
**kobalicek (Petr Kobalicek)**
    to me it seemed that emscripten would just create one "statically" linked wasm archive
**iFire**
    it would be similar to a shared library?
**kobalicek (Petr Kobalicek)**
    I mean the way asmjit works, you can generate a single function, for example, I'm not sure how this would work in wasm - because when running asmjit natively, that generated function can access memory of the process - if you create a wasm function, would it have access to wasm code loaded by javascript, for example?
    What I mean is, how would that wasm code even reference another wasm code?
    is it possible to free that wasm code, etc
    because if you think of it, when running natively, you can do all sorts of stuff, allocate executable memory, fill it with code, execute it, change it, free it, etc
     have no idea about how restricted wasm is to be honest
    not sure it would even be complicated to generate it, I think the tooling around it would be more difficult than generating the bytecode

**Comment by AdamFrisby** [comment]
It's possible yeah, just be aware that sandboxing any code that way might be a painful exercise (if you need the sub wasm running at a lower privilege level with discrete memory)

One route if you do need that sandboxing would be to emscripten compile something like wasmtime. Otherwise you can load the modules into the main space at runtime easily enough.
________________________________
From: K. S. Ernest (iFire) Lee ***@***.***>
Sent: Saturday, January 8, 2022 7:07:22 AM
To: omigroup/omigroup ***@***.***>
Cc: Adam Frisby ***@***.***>; Mention ***@***.***>
Subject: [omigroup/omigroup] WASM - Godot Engine scripting (Discussion #160)


@AdamFrisby<https://github.com/AdamFrisby>

Godot Engine works on 5-10 platforms and has a scripting language called gdscript. One of the platforms is WASM. I wanted to investigate using jitasm for gdscript.

Since I wanted to try using gdscript for wasm, as a sandboxing layer, here are some questions.

Can we generate wasm in wasm and have the wasm called?

Here's a chat log with the asmjit community:

kobalicek: (Petr Kobalicek)
you would be basically generating wasm code from wasm code, and wanting wasm code to call that generated wasm code
I have no idea whether it's possible or not
iFire
I have a feeling it is, but I'm new to this.
kobalicek (Petr Kobalicek)
to me it seemed that emscripten would just create one "statically" linked wasm archive
iFire
it would be similar to a shared library?
kobalicek (Petr Kobalicek)
I mean the way asmjit works, you can generate a single function, for example, I'm not sure how this would work in wasm - because when running asmjit natively, that generated function can access memory of the process - if you create a wasm function, would it have access to wasm code loaded by javascript, for example?
What I mean is, how would that wasm code even reference another wasm code?
is it possible to free that wasm code, etc
because if you think of it, when running natively, you can do all sorts of stuff, allocate executable memory, fill it with code, execute it, change it, free it, etc
because if you think of it, when running natively, you can do all sorts of stuff, allocate executable memory, fill it with code, execute it, change it, free it, etc
have no idea about how restricted wasm is to be honest
not sure it would even be complicated to generate it, I think the tooling around it would be more difficult than generating the bytecode

—
Reply to this email directly, view it on GitHub<https://github.com/omigroup/omigroup/discussions/160>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AAA326JVK5EYGTNVKGAR5EDUU5BXVANCNFSM5LPRM5FA>.
Triage notifications on the go with GitHub Mobile for iOS<https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android<https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.

**Comment by Toasterson** [comment]
WASM at the end of the day is just bytecode like the native gdscript. Sandboxing it inside the engine specially can be tricky but is mainly a question about the ABI. so something like WASMER is good to extend godot with a WASM sandbox. but keep in mind it will still be able to do everything inside the engine like any other script. Godot has no Node security system.

**Comment by AdamFrisby** [comment]
The issue with sandboxing is this is UGC, where content is likely loaded at runtime from untrusted sources. Some level of sandboxing should be considered essential, but you're right - wasmer et al can handle that job.

Something to discuss on the meeting next week.
________________________________
